### PR TITLE
Unify encoding to 'utf-8' and optimize error rate calculation.

### PR DESCRIPTION
**Pull request by pkuyym**
1. Use `codecs` to make sure that all texts are 'utf-8' encoding including vocabulary and transcriptions. 
2. Optimize error rate calculation. Mainly reduce space requirement from O(m*n) to O(min(m, n)).

### CODE REVIEW

**Review by wangkuiyi — COMMENTED** [review]
LGTM!

赞细致的工作风格！

File: deep_speech_2/datasets/librispeech/librispeech.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import soundfile
 import json
 from paddle.v2.dataset.common import md5file
+import codecs
```
Comment (xinghai-sun) [inline deep_speech_2/datasets/librispeech/librispeech.py]: Put `import codecs` before `from paddle.v2.dataset.common import md5file`. 
Please keep the import order according to [Google Coding Style for Python](https://google.github.io/styleguide/pyguide.html?showone=Imports_formatting#Imports_formatting):

- standard library imports
- third-party imports
- application-specific imports

File: deep_speech_2/data_utils/featurizer/text_featurizer.py [hunk 2]
```diff
@@ -59,7 +60,7 @@ def _char_tokenize(self, text):
     def _load_vocabulary_from_file(self, vocab_filepath):
         """Load vocabulary from file."""
         vocab_lines = []
-        with open(vocab_filepath, 'r') as file:
+        with codecs.open(vocab_filepath, 'r', 'utf-8') as file:
```
Comment (xinghai-sun) [inline deep_speech_2/data_utils/featurizer/text_featurizer.py:63]: What happens if the vocab_filepath is not in utf-8 format ? Any exception thrown out to tell users about it?

File: deep_speech_2/error_rate.py [hunk 3]
```diff
@@ -10,47 +10,52 @@
...
+    """Levenshtein distance is a string metric for measuring the difference
+    between two sequences. Informally, the levenshtein disctance is defined as
+    the minimum number of single-character edits (substitutions, insertions or
+    deletions) required to change one word into the other. We can naturally
+    extend the edits to word level when calculate levenshtein disctance for
+    two sentences.
     """
-    ref_len = len(ref)
-    hyp_len = len(hyp)
+    m = len(ref)
```
Comment (xinghai-sun) [inline deep_speech_2/error_rate.py:20]: Why to replace `ref_len` with `m`? `ref_len` seems to be more readable.

File: deep_speech_2/error_rate.py [hunk 4]
```diff
@@ -10,47 +10,52 @@
...
+    for j in xrange(n + 1):
         distance[0][j] = j
-    for i in xrange(ref_len + 1):
-        distance[i][0] = i
 
     # calculate levenshtein distance
-    for i in xrange(1, ref_len + 1):
-        for j in xrange(1, hyp_len + 1):
+    for i in xrange(1, m + 1):
+        distance[i % 2][0] = i
```
Comment (luotao1) [inline deep_speech_2/error_rate.py]: 可以用一个临时变量存 i%2和(i-1)%2。省去每轮多算的n+1次。

File: deep_speech_2/tests/test_error_rate.py [hunk 5]
```diff
@@ -34,21 +34,33 @@ def test_cer_1(self):
 
     def test_cer_2(self):
         ref = 'werewolf'
+        hyp = 'weae  wolf'
+        char_error_rate = error_rate.cer(ref, hyp, remove_space=True)
```
Comment (xinghai-sun) [inline deep_speech_2/tests/test_error_rate.py:76]: 1. Add a test case: `'were wolf'` + `'were  (multiple whitespace)   wolf'`, cer=0,  for remove_space=False.
2. Could you please add more test case for wer()? Currently we have only one, which is not safe.

**Review by xinghai-sun — CHANGES_REQUESTED** [review]

File: deep_speech_2/tests/test_error_rate.py (hunk 5 above)
Comment (pkuyym) [reply deep_speech_2/tests/test_error_rate.py:76]: Done.

File: deep_speech_2/error_rate.py (hunk 4 above)
Comment (pkuyym) [reply deep_speech_2/error_rate.py]: Done.

File: deep_speech_2/error_rate.py (hunk 3 above)
Comment (pkuyym) [reply deep_speech_2/error_rate.py:20]: ```m```, ```n``` indicate the row number and column number for DP.

File: deep_speech_2/datasets/librispeech/librispeech.py (hunk 1 above)
Comment (pkuyym) [reply deep_speech_2/datasets/librispeech/librispeech.py]: Done.

File: deep_speech_2/data_utils/featurizer/text_featurizer.py (hunk 2 above)
Comment (pkuyym) [reply deep_speech_2/data_utils/featurizer/text_featurizer.py:63]: Yes, exception like ```IOException``` and ```UnicodeDecodeError``` will be throwed up when reading invalid files.

**Review by kuke — COMMENTED** [review]
LGTM

**Review by xinghai-sun — APPROVED** [review]
LGTM.
